compiler/passes/denormalize.cpp

```diff
@@ -669,3 +676,259 @@ inline bool unsafeExprInBetween(Expr* e1, Expr* e2, Expr* exprToMove,
   }
   return false;
 }
+
+/////////////////////////////////////////////////////////////////////////////
+// undoReturnByRef()
+//
+// Ultimately we want to avoid introducing return by ref.
+// For now, simply undo it as a denormalization step.
+
+///////////
+//
+// transformRetTempDef() will modify 'fn' to replace
+//   =(refArg, something)
+//   return(void)
+// with:
+//   return(something)
+
+class DefInfoRR {
+public:
+  FnSymbol*  fn;           // the function that returns by ref
+  ArgSymbol* refArg;       // the function's formal implementing that
+  bool       skipLnFnArgs; // if true, refArg is fn's 3rd-last formal
+};
+
+static bool isGoodRefArg(FnSymbol* fn,              DefInfoRR& info,
+                         bool      skippedLnFnArgs, Expr*      argDef) {
+  ArgSymbol* refArg = toArgSymbol(toDefExpr(argDef)->sym);
+  if (!refArg->hasFlag(FLAG_RETARG))
+    return false; // give up; maybe look harder?
+  INT_ASSERT(!strcmp(refArg->name, "_retArg"));
+  INT_ASSERT(fn->retType = dtVoid);
+
+  info = { fn, refArg, skippedLnFnArgs };
+  return true;
+}
+
+static bool acceptableDef(FnSymbol* fn, DefInfoRR& info) {
```

Comment:
Maybe we could make this into a method `acceptable` in the `UseInfo` type? And then the `UseInfo` constructor can take the three arguments it needs, or you can write a new one if you need to compute some things. We can do the same thing for `UseInfo`.

Then later patterns can be written like:

```cpp
UseInfo info(foo, bar, baz);
if (!info.acceptable()) continue;
```